### PR TITLE
Add settings for keeper retries during INSERT

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Documentation (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Settings for Keeper retries during INSERT into replicated MergeTree

### CODE REVIEW

File: docs/en/operations/settings/settings.md [hunk 1]
```diff
@@ -1632,6 +1632,49 @@ SELECT * FROM test_table
 └───┘
 ```
 
+## insert_keeper_max_retries
+
+The setting sets the maximum number of retries for ClickHouse Keeper (or ZooKeeper) requests during insert into replicated MergeTree. Only Keeper requests which failed due to network error, Keeper session timeout or request timeout considered for retries.
```
Comment (DanRoscigno) [inline docs/en/operations/settings/settings.md:1637]:
```suggestion
The setting sets the maximum number of retries for ClickHouse Keeper (or ZooKeeper) requests during insert into replicated MergeTree. Only Keeper requests which failed due to network error, Keeper session timeout, or request timeout are considered for retries.
```

File: docs/en/operations/settings/settings.md [hunk 2]
```diff
@@ -1632,6 +1632,49 @@ SELECT * FROM test_table
...
+
+Possible values:
+
+-   Positive integer.
+-   0 — Retries are disabled
+
+Default value: 0
+
+Keeper request retries are done after some timeout. The timeout is controlled by the following settings: `insert_keeper_retry_initial_backoff_ms`, `insert_keeper_retry_max_backoff_ms`.
+The first retry is done after `insert_keeper_retry_initial_backoff_ms` timeout. The consequent timeouts will be calculated as following:
```
Comment (DanRoscigno) [inline docs/en/operations/settings/settings.md:1647]:
```suggestion
The first retry is done after `insert_keeper_retry_initial_backoff_ms` timeout. The consequent timeouts will be calculated as follows:
```

File: docs/en/operations/settings/settings.md [hunk 3]
```diff
@@ -1632,6 +1632,49 @@ SELECT * FROM test_table
...
+timeout = min(insert_keeper_retry_max_backoff_ms, latest_timeout * 2)
+```
+
+For example, if `insert_keeper_retry_initial_backoff_ms=100`, `insert_keeper_retry_max_backoff_ms=10000` and `insert_keeper_max_retries=8` then timeouts will be `100, 200, 400, 800, 1600, 3200, 6400, 10000`.
+
+Apart from fault tolerance, the retries aims to provide better user experience, - they allow to avoid returning a user an error during INSERT execution if Keeper is restarted, for example, due to upgrade.
+
+## insert_keeper_retry_initial_backoff_ms {#insert_keeper_retry_initial_backoff_ms}
+
+Initial timeout(in milliseconds) to retry failed Keeper request during INSERT query execution
```
Comment (DanRoscigno) [inline docs/en/operations/settings/settings.md:1658]:
```suggestion
Initial timeout(in milliseconds) to retry a failed Keeper request during INSERT query execution
```

File: docs/en/operations/settings/settings.md [hunk 4]
```diff
@@ -1632,6 +1632,49 @@ SELECT * FROM test_table
...
+Possible values:
+
+-   Positive integer.
+-   0 — No timeout
+
+Default value: 100
+
+## insert_keeper_retry_max_backoff_ms {#insert_keeper_retry_max_backoff_ms}
+
+Maximum timeout(in milliseconds) to retry failed Keeper request during INSERT into Rquery execution
```
Comment (DanRoscigno) [inline docs/en/operations/settings/settings.md:1669]:
```suggestion
Maximum timeout (in milliseconds) to retry failed Keeper requests during INSERT query execution
```

**Review by DanRoscigno — CHANGES_REQUESTED** [review]

**Comment by DanRoscigno** [comment]
Thanks so much @devcrafter !

File: docs/en/operations/settings/settings.md [hunk 5]
```diff
@@ -2334,7 +2377,8 @@ Default value: 8.
 
 ## background_schedule_pool_size {#background_schedule_pool_size}
 
-Sets the number of threads performing background tasks for [replicated](../../engines/table-engines/mergetree-family/replication.md) tables, [Kafka](../../engines/table-engines/integrations/kafka.md) streaming, [DNS cache updates](../../operations/server-configuration-parameters/settings.md/#server-settings-dns-cache-update-period). This setting is applied at ClickHouse server start and can’t be changed in a user session.
+Sets the number of threads performing background tasks for [replicated](../../engines/table-engines/mergetree-family/replication.md) tables, [Kafka](../../engines/table-engines/integrations/kafka.md) streaming, [DNS cache updates](../../operations/server-configuration-parameters/
```
Comment (devcrafter) [inline docs/en/operations/settings/settings.md:2380]: Not sure why this change is here

Comment (DanRoscigno) [reply docs/en/operations/settings/settings.md:2380]: I will checkout the PR and see what happened.

**Review by DanRoscigno — APPROVED** [review]
LGTM

File: docs/en/operations/settings/settings.md (hunk 5 above)
Comment (DanRoscigno) [reply docs/en/operations/settings/settings.md]: LGTM again ;)
